src/plugins/graphql.js

```diff
@@ -2,6 +2,43 @@
 
 const platform = require('../platform')
 
+// Because we want to use the parsed operation name as the span's resource name, we can't really start the span just
+// yet. Instead we store the actual start time now and use that at the time of creating the operation span.
+function createWrapGraphQL (tracer, config) {
+  return function wrapGraphQL (graphql) {
+    return function graphqlWithTrace (
+      argsOrSchema,
+      source,
+      rootValue,
+      contextValue,
+      variableValues,
+      operationName,
+      fieldResolver
+    ) {
+      if (arguments.length === 1) {
+        source = argsOrSchema.source
+        rootValue = argsOrSchema.rootValue
+        contextValue = argsOrSchema.contextValue
+        variableValues = argsOrSchema.variableValues
+        operationName = argsOrSchema.operationName
+        fieldResolver = argsOrSchema.fieldResolver
+        argsOrSchema = argsOrSchema.schema
+      }
+      if (!contextValue) {
+        contextValue = {}
+      }
+      if (!contextValue._datadog_operation_start_time) {
+        Object.defineProperties(contextValue, {
+          _datadog_operation_start_time: {
```

Comment:
Instrumenting at this layer is not safe as libraries such as `express-graphql` skip it and use parse and execute directly. Maybe this check can be replaced by middleware instrumentation on the web server instead? The timing of any GraphQL middleware should roughly match the timing of the entire operation. It would however mean that if used without a parent, `parse`, `validate` and `query` would end up as different traces.

Reply:
Actually, maybe this could start some kind of `graphql.operation` span that would be the parent when available. Then it would be possible to start a span directly in `parse` and `validate` instead of deferring them. When used with wrappers such as `express-graphql` the `graphql.operation` span would be skipped since this file is not used, but then you have the parent from express.

Do you think that would make sense?

Reply:
> When used with wrappers such as `express-graphql` the `graphql.operation` span would be skipped since this file is not used, but then you have the parent from express.

In my naive thoughts it seems like that may not provide the greatest experience when drilling down into the data. Ideally we do want the named operation to be the parent of these, no?

Is it possible for this plugin to optionally hook into another package? i.e. _if_ the `express-graphql` package is installed wrap `graphqlHTTP` from it and then do the same recording of `_datadog_operation_start_time`?

Reply:
Thinking about it some more, maybe we can just, in order, check for the existence of `document._datadog_parse_time` and then `document._datadog_validate_time`. If either of those exists we use their start time as the start time of the operation span. That way we don’t need to hook either the `graphql` function or anything in `express-graphql`. What do you think?

Reply:
That would work. I would still prefer to distinguish between the entire operation and the execution as 2 different spans though. So basically there would be a root operation span, with `parse`, `validate` and `execute` each having their own spans with the operation as the parent.

The only problem I see with this is that if `parse` or `validate` fails, then `execute` is never called and you would not get a trace at all. I don't know however if there is a use case to get a trace in that case, especially since if there is a parent, the parse/validate error could be reported in the `error.*` tags of the parent.

Reply:
Makes sense. The only problem I have is what to ‘name’ the root span. Currently the span gets the name of the operation, but that requires parsing first. We can lazily create the root span in parse, validate, and execute and use the parsed operation name, but what name do we use when parsing fails?

Also, I assume you’d want to attach the full document and variables to the root span?